File: using-python/abstract_factory_pattern.py
from abc import ABC, abstractmethod

# Abstract factory interface
class GUIFactory(ABC):
    
    @abstractmethod
    def create_button(self):
        pass

    @abstractmethod
    def create_checkbox(self):
        pass


class Button(ABC):
    @abstractmethod
    def paint(self):
        pass


class Checkbox(ABC):
    @abstractmethod
    def paint(self):
        pass


class WinButton(Button):
    def paint(self):
        print("Render a button in Windows style.")

class MacButton(Button):
    def paint(self):
        print("Render a button in Mac style.")

class WinCheckbox(Checkbox):
    def paint(self):
        print("Render a checkbox in Windows style.")


class MacCheckbox(Checkbox):
    def paint(self):
        print("Render a checkbox in Mac style.")


# Windows factory implmentation (Concrete)
class WinFactory(GUIFactory):
    def create_button(self):
        return WinButton()
    
    def create_checkbox(self):
        return WinCheckbox()
    

class MacFactory(GUIFactory):
    def create_button(self):
        return MacButton()
    
    def create_checkbox(self):
        return MacCheckbox()
    

class Application:
    def __init__(self, factory: GUIFactory) -> None:
            self.factory = factory
            self.button = None

    def create_ui(self):
        self.button = self.factory.create_button()

    def paint(self):
        self.button.paint()


class ApplicationConfigurator:

    def read_application_config_file(self):
        return {'OS': 'Windows'}

    def main(self):
        config = self.read_application_config_file()

        if config['OS'] == "Windows":
            factory = WinFactory()
        elif config['OS'] == "Mac":
            factory = MacFactory()
        else: 
            raise Exception("Error! Unknown operating system.")
        
        app = Application(factory)
        app.create_ui()
        app.paint()


configurator = ApplicationConfigurator()
configurator.main()

File: using-typescript/abstract-factory-pattern.ts
interface Button {
    paint(): void;
}

interface Checkbox {
    paint(): void;
}

interface GUIFactory {
    createButton(): Button;
    createCheckbox(): Checkbox;
}

class WinButton implements Button {
    paint(): void {
        console.info("Render a button in Windows style.")
    }
}

class WinCheckbox implements Checkbox {
    paint(): void {
        console.info("Render a checkbox in Windows style.")
    }
}

class MacButton implements Button {
    paint(): void {
        console.info("Render a button in Mac style.")
    }
}

class MacCheckbox implements Checkbox {
    paint(): void {
        console.info("Render a checkbox in Mac style.")
    }
}

class WinFactory implements GUIFactory {
    createButton(): Button {
        return new WinButton();
    }

    createCheckbox(): Checkbox {
        return new WinCheckbox();
    }
}

class MacFactory implements GUIFactory {
    createButton(): Button {
        return new MacButton();
    }

    createCheckbox(): Checkbox {
        return new MacCheckbox()
    }
}

class Application {
    private factory: GUIFactory;
    private button: Button | null;

    constructor(factory: GUIFactory) {
        this.factory = factory;
        this.button = null;
    }

    createUI(): void {
        this.button = this.factory.createButton();
    }

    paint(): void {
        if (this.button) {
            this.button.paint();
        }
    }
}

class ApplicationConfigurator {
    readApplicationConfigFile(): { OS: string } {
        return { OS: 'Windows' };
    }

    main(): void {
        const config = this.readApplicationConfigFile();
        let factory: GUIFactory;

        if (config.OS === "Windows") {
            factory = new WinFactory();
        } else if (config.OS === "Mac") {
            factory = new MacFactory()
        } else {
            throw new Error("Error! Unknows operating system.")
        }

        const app = new Application(factory);
        app.createUI();
        app.paint();
    }
}

const configurator = new ApplicationConfigurator();
configurator.main();